File: 079WordSearch/WordSearch.py
# -*- coding: utf -8-*-

from typing import List

class WordSearch:
    def exist(self, board: "List[List[str]]",
              word: "str") -> "bool":
        '''
        over time limit, fuck it
        '''
        rows = len(board)
        if not rows: return False
        cols = len(board[0])
        used = [[False] * cols for _ in range(rows)]
        for i in range(rows):
            for j in range(cols):
                if self.__exist_(board, word, i, j,
                                 0, used): return True
        return False

    def __exist_(self, board: "List[List[str]]", word: "str",
                 row: "int", col: "int", index: "int",
                 used: "List[List[bool]]") -> "bool":
        ## out of bound
        if row not in range(0, len(board)) or\
              col not in range(0, len(board[0])): return False
        ## used
        if used[row][col]: return False
        ## not what we are looking for
        if board[row][col] != word[index]: return False
        ## found the entire thing
        if index == len(word) - 1: return True
        used[row][col] = True
        ## try four directions
        ## up
        if self.__exist_(board, word, row - 1, col,
                         index + 1, used): return True
        ## down
        if self.__exist_(board, word, row + 1, col,
                         index + 1, used): return True
        ## left
        if self.__exist_(board, word, row, col - 1,
                         index + 1, used): return True
        ## right
        if self.__exist_(board, word, row, col + 1,
                         index + 1, used): return True
        ## nothing works
        used[row][col] = False
        return False


def main():
    test = WordSearch()
    print(test.exist([
        ['A', 'B', 'C', 'E'],
        ['S', 'F', 'C', 'S'],
        ['A', 'D', 'E', 'E']
    ], "ABCCED"))
    print(test.exist([
        ['A', 'B', 'C', 'E'],
        ['S', 'F', 'C', 'S'],
        ['A', 'D', 'E', 'E']
    ], "SEE"))
    print(test.exist([
        ['A', 'B', 'C', 'E'],
        ['S', 'F', 'C', 'S'],
        ['A', 'D', 'E', 'E']
    ], "ABCB"))

if __name__ == "__main__":
    main()

File: 079WordSearch/WordSearch.cpp


#include <vector>
#include <string>
#include <iostream>

class WordSearch {
public:
    bool exist(std::vector<std::vector<char> >& board, std::string word) {
        // visited
        // rows of visited
        // std::vector<bool> row;
        size_t m = board.size();
        if (m == 0) return false;
        // for (size_t i = 0; i < board[0].size(); ++i) {
        //     row.push_back(false);
        // }
        size_t n = board[0].size();
        std::vector<std::vector<bool> > visited(m, std::vector<bool>(n, false));
        // for (size_t i = 0; i < m; ++i) {
        //     visited.push_back(row);
        // }
        // visited[0][0] = true;
        for (size_t i = 0; i < m; ++i) {
            for (size_t j = 0; j < n; ++j) {
                // std::cout << visited[i][j];
                // if (j != board[0].size() - 1) {
                //     std::cout << ", ";
                // }
                if (exist_(board, word, visited, 0, i, j)) {
                    return true;
                }
            }
            // std::cout << std::endl;
        }
        return false;
    }

private:
    bool exist_(
        std::vector<std::vector<char> >& board,
        std::string& word,
        std::vector<std::vector<bool> >& visited,
        size_t idx,
        size_t r,
        size_t c
    ) {
        // out of bound
        // if ((r >= board.size() || r < 0) && (c >= board[0].size() || c < 0)) {
        if (r >= board.size() || r < 0 || c >= board[0].size() || c < 0) {
            return false;
        }
        // visited or wrong char
        if (visited[r][c] || board[r][c] != word.at(idx)) {
            return false;
        }
        // got all letters
        if (idx == word.size() - 1) return true;
        // check four directions
        std::pair<int, int> dirs[4] = {
            std::pair<int, int>(1, 0),
            std::pair<int, int>(-1, 0),
            std::pair<int, int>(0, 1),
            std::pair<int, int>(0, -1),
        };
        visited[r][c] = true;
        for (size_t i = 0; i < 4; ++i) {
            if (
                exist_(
                    board,
                    word,
                    visited,
                    idx + 1,
                    r + dirs[i].first,
                    c + dirs[i].second
                )
            ) {
                return true;
            }
        }
        // nothing from any of the four dirs
        visited[r][c] = false;
        return false;
    }
};


int main(void) {
    WordSearch test;
    std::vector<std::vector<char> > test1(3, std::vector<char>(5, 'a'));
    std::cout << test.exist(test1, "abcde") << std::endl;
}
